### DISCUSSION TITLE
Multi value wasm, how does it works?

### DISCUSSION THREAD

**Discussion by msqr1** [emscripten-core/emscripten#21792] [Q&A] [answered]
The first idea that comes to my mind when we have multi value wasm is that we just break down struct in memory until it contains just primitive datatype in the correct order. Then we just return or store the return value in correct or with the correct offset in memory. Is that true how it works?

**Comment by sbc100** [comment, marked as answer]
Basically yes, although exactly how and when it makes sense to do this kind of thing is up in the air still.   This is really question for upstream llvm which does the code generation.   There is an experiment-mv ABI there that is not currently well supported but you can play with it you like.   On the emscripten side will likely not make any changes to the ABI aside from following whatever upstream llvm does.

Separately to the C ABI there is talk of enabling multi-value as a feature so we can use it internally for things (e.g. using assembly perhaps).    

For recent work on this see: https://github.com/llvm/llvm-project/pull/88492
And https://github.com/WebAssembly/tool-conventions/pull/223
And https://github.com/llvm/llvm-project/pull/82714

**Comment by msqr1** [comment]
Why not do this, for now, for every struct that is breakable into basic wasm types? Maybe I'm not experienced enough to see the issue?

**Reply by sbc100** [reply]
What you are suggesting is essentially what the `experimental-mv` ABI in LLVM does.   However, when we were doing experiments with this a while back it didn't prove for be advantageous, and in particular i believe it make some optimization in binaryen more difficult, and actually resulting in worse code size and/or performance.    @tlively can speak more to this perhaps.

**Reply by msqr1** [reply]
What is the current way when we are returning struct if not this? And what kind of optimization can't be done? @sbc100

**Reply by tlively** [reply]
The problem is that if you try to return a struct with hundreds or thousands of fields via multivalue, then code size bloat from getting all those values onto the Wasm stack in the right order and then consuming them all afterward would be huge. The best solution would be to pass structs up to a certain size using multivalue return, then fall back to the current method, which is to modify the function signature so it takes a pointer into memory where the returned struct should be written to.

**Reply by msqr1** [reply]
Why don't we just have struct definition for wasm and still restrict to single retval instead of multiple retval? We can define it once and reuse it?

**Reply by tlively** [reply]
Yes, that's actually what happens today. Even with multivalue enabled, we still always pass structs via memory and never have more than one return value. The exception is when you enable the `experimental-mv` ABI that @sbc100 mentioned.

**Reply by msqr1** [reply]
No, what I mean is it would be nice to have in the wasm spec struct instead of just primitive types

**Reply by tlively** [reply]
Oh, I see. We actually are adding struct types in [WasmGC](https://github.com/WebAssembly/gc), but they're allocated on a heap separate from the Wasm memory C++ uses. Defining a structured view over linear memory would most likely be considered too high-level a feature to be added to the Wasm spec.

**Reply by msqr1** [reply]
What is the difference between that struct and view over memory?

**Reply by tlively** [reply]
WasmGC structs do not live in the linear memory. They cannot be accessed with normal load and store instructions and it is not possible to have a pointer to one of their fields.

**Reply by msqr1** [reply]
What is the new changes in 3.1.59 that enables this by default. What exactly does "This only enables the language features and does not turn on the multivalue ABI"? What is the difference between the language features and the ABI?

**Reply by tlively** [reply]
Enabling the multivalue feature means that it is valid for optimizations to produce new uses of multivalue (for example by using multivalue blocks inside a function or multivalue signatures if splitting up a function), but it does not affect the ABI or calling convention of the functions clang emits.

**Reply by sbc100** [reply]
It also means you want write your own multi-value code using inlines assembly (although I'm not sure it its support in the current .s format) or wat and link that into the emscripten program.

**Reply by msqr1** [reply]
So this doesn't let me return struct with fields to explicitly target multivalue, but the optimization can do it. If I want it though, I can still do inline asm?

**Reply by sbc100** [reply]
You can create your own functions that pass and return multi-value (e.g. with some custom compiler), but clang won't do that by default.   That would require a change in clang C ABI.  There is an experimental clang ABI flag you can pass if you are interesting in playing with it, but emscripten doesn't plan on using anything like by default any time soon.

**Reply by msqr1** [reply]
So it's is valid to generate new uses of multivalue (not restricted), but clang won't do it anytime soon?

**Reply by sbc100** [reply]
Exactly
